doc/cmml-timeshift.1
.\" $Header: /aolnet/dev/src/CVS/sgml/docbook-to-man/cmd/docbook-to-man.sh,v 1.1.1.1 1998/11/13 21:31:59 db3l Exp $
.\"
.\"	transcript compatibility for postscript use.
.\"
.\"	synopsis:  .P! <file.ps>
.\"
.de P!
.fl
\!!1 setgray
.fl
\\&.\"
.fl
\!!0 setgray
.fl			\" force out current output buffer
\!!save /psv exch def currentpoint translate 0 0 moveto
\!!/showpage{}def
.fl			\" prolog
.sy sed -e 's/^/!/' \\$1\" bring in postscript file
\!!psv restore
.
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
'\" t 
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n  
.TH "cmml-timeshift" "1" 
.SH "NAME" 
cmml-timeshift \(em       Apply a timeshift to all the clip tags. 
 
.SH "SYNOPSIS" 
.PP 
\fBcmml-timeshift\fR [-o \fBfile\fR  | --output \fBfile\fR ]  [-s \fBoffset\fR  | --sec \fBoffset\fR ]  [-h  | --help ]  [-v  | --version ] filename  
.SH "Description" 
.PP 
\fBcmml-timeshift\fR parses a CMML instance document 
given as \fBfilename\fR, applies a timeshift 
to the start and end times of the clip tags and creates a 
CMML instance document from it, which gets written to stdout 
by default. Clip tags that would receive negative start times 
are skipped. In case of a non-recoverable error upon parsing, the 
faulty tag including line and column number is reported. 
 
.SH "Options" 
.PP 
\fBcmml-timeshift\fR accepts the following options: 
 
.IP "-o \fBfile\fR, --output \fBfile\fR" 10 
Output file. Writes fixed CMML document to the 
given \fBfile\fR instead of stdout, which is the 
default. 
.IP "-s \fBoffset\fR, --sec \fBoffset\fR" 10 
Timeshift offset in seconds. Shifts the start- and 
endtime of the clip tags by \fBoffset\fR seconds. 
 
.IP "-h, --help" 10 
Print usage message. 
.IP "-v, --version" 10 
Print version information. 
.SH "VERSION" 
.PP 
This man page is correct for CMML version 2.1 and libcmml 
version 0.9.0. 
 
.SH "SEE ALSO" 
.PP 
More information is available at: 
\fIAnnodex.net (link to URL http://www.annodex.net/) \fR    
.\" created by instant / docbook-to-man, Mon 18 Apr 2005, 11:26 


ChangeLog
2008-04-05 Silvia Pfeiffer <silvia at silvia-pfeiffer dot de>
    * Version 0.9.4
    * added automake 1.10
	* check differences from cmml-validate's stdout and stderr separately in tests/Makefile.am, as the two could be arbitrarily interleaved if 2>&1 is used.
	* fixed make distcheck

2006-03-10  Silvia Pfeiffer <silvia at silvia-pfeiffer dot de>
	* Version 0.9.2
	* changed "class" struct member names to make libcmml compatible with C++
	* this also required a change of the SHARED_VERSION_INFO
	
2006-03-10  Silvia Pfeiffer <silvia at silvia-pfeiffer dot de>
	* removed executable bits from examples/cmml-write.c
	* also includes changes from Alex Krumm-Heller to fix
	  building the library on Windows

2006-01-31  Silvia Pfeiffer <silvia at silvia-pfeiffer dot de>
	* Version 0.9.1
	* nothing new: just added Thomas' build fixes
	
2005-06-13  Thomas Vander Stichele  <thomas at apestaart dot org>

	* configure.ac:
	  uniformize configure's output across the annodex stack

2005-06-13  Thomas Vander Stichele  <thomas at apestaart dot org>

	* doc/Makefile.am:
	  the doxygen stamp file is a build marker and should not be installed

2005-06-07  Thomas Vander Stichele  <thomas at apestaart dot org>

	* m4/as-ac-expand.m4:
	  add an m4
	* autogen.sh:
	  add m4 to aclocal
	* doc/Makefile.am:
	  use datadir instead of prefix/data
	* configure.ac:
	  fix up the locations of where configure thinks stuff gets installed

2005-04-18  Silvia Pfeiffer <silvia.pfeiffer@csiro.au>
	* Version 0.9.0
	* support for CMML version 2.1
	* added cmml-timeshift tool

2005-03-31  Silvia Pfeiffer <silvia.pfeiffer@csiro.au>
	* Version 0.8.3
	* fixed cmml-fix's warning reporting at EOF
	* fixed make check
	* fixed npt parsing

2004-03-11  Silvia Pfeiffer <silvia.pfeiffer@csiro.au>
	* Version 0.8.2
	* took CVS out of tarball
	* little fixes

2003-12-31  Zen Kavanagh <Zen@illiminable.com>
	* Version 0.8.1
	* fixed Windows port ('b' flag into fopens, getopt)

2003-12-06  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.8.0
	* adapted to CMML 2.0
	* fixed up much error handling, e.g. no stream tag,
	  head or anchor before stream, missing required attributes,
	  param outside source tag, track default value
	* created an error model of sloppy and non-sloppy errors
	* extended documentation, all in doxygen
	* included manpages
	* change from file descriptors to FILE *, changing API 
	  functions cmml_new & cmml_destroy
	* added cmml-fortune (contributed by Conrad Parker)
	* added cmml-fix (thanks to Conrad Parker)
	* included getot_long for cmml-validate, cmml-fix, cmml-fortune
	* included stdin for cmml-validate, cmml-fix, cmml-fortune
	* added several API functions for preamble handling

2003-10-21  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.7.7
	* fixed up character escaping when printing attribute and 
	  element values

2003-10-19  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.7.6
	* fixed up example code and documentation
	* included Version_script
	* included some more testing using the example programs
	* fixed some bugs in cmml.c on seeking and added a function

2003-10-18  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.7.5
	* fixed up "make distcheck" and also added "make test"
	* included Doxygen documentation
	* fixed some memory holes in cmml_element.c

2003-09-25  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.7.4
	* a negative timebase gets now flagged as a proper error

2003-09-23  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.7.3
	* fixed a bug for anchors without start time

2003-09-22  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.7.2
	* adapted time specification to "npt:12.5" from "npt=12.5" etc.

2003-09-15  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.7.0
	* windows port (subdirectory Win32)

2003-06-22  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.8
	* bug fixes & added ignoring of multiple stream & head tags
	  into cmml-transform

2003-06-22  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.7
	* changed seeking structure, deleted "seek" functions, added
	  cmml_set and cmml_skip_to functions

2003-06-21  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.6
	* added support for seeking ranges

2003-06-20  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.5
	* added tool to demonstrate a server's implementation of cmml parsing
	* bug fixes

2003-06-01  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.4
	* added a "end" attribute to the "media" tag in "stream"

2003-06-01  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.3
	* deleted all amml support
	* deleted "seekoffset" and "seekend" attribute
	  in "media" tag as the "location" attribute can handle them

2003-06-01  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.2
	* added "seekend" attribute to "media" tag

2003-05-27  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.1
	* added a test file for amml testing
	* fixed up the dtds

2003-05-25  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.6.0 contains major changes in the cmml
	* supporting multiple mimetype input bitstream multiplexing
	* supporting variable bitrate in ANNODEX
	* supporting several more smpte specifications

2003-05-25  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.5.10
	* adds a cmml-transcode-dtd function to transcode cmml files
	  from the dtd of version 0.5.9 to the one in 0.6.10; this
	  function is deleted in 0.6.10 again

2003-05-21  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.5.9
	* fixed up utc support in head, anchors and for seeking

2003-04-05  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.5.8
	* supports changed media mapping, i.e. deleted packetrate
	* corrected time support to use absolute times in attributes
	  instead of relative ones; deleted support for start/end/now

2003-03-21  Silvia Pfeiffer <Silvia.Pfeiffer@csiro.au>
	* Version 0.5.6 includes initial utc support


tools/cmml-timeshift.c
/* Copyright (C) 2005 CSIRO Australia

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:
   
   - Redistributions of source code must retain the above copyright
   notice, this list of conditions and the following disclaimer.
   
   - Redistributions in binary form must reproduce the above copyright
   notice, this list of conditions and the following disclaimer in the
   documentation and/or other materials provided with the distribution.
   
   - Neither the name of the CSIRO nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.
   
   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
   PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE ORGANISATION OR
   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
   EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
   PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <fcntl.h>
#include <errno.h>

#ifndef WIN32
#include <unistd.h>
#endif

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#include <cmml.h>

/**
 * \file
 * cmml-timeshift:
 * parses a CMML instance document and applies a timeshift to
 * all the clip tags.
 *
\verbatim
Usage: cmml-timeshift [options] filename
Apply a timeshift to all the clip tags.
 
Possible options:
  -s offset, --sec offset
                 Specify the offset in seconds to add to clip tags
  -o filename, --output filename
                 Specify the output filename. The file is written
                 to standard output by default.
  -h, --help     Display this help information
  -v, --version  Display version information
\endverbatim
 *
 */

/** the size of the print buffer */
#define BUFSIZE 100000

/*
#define DEBUG
*/

/**
 * outfile: defines FILE pointer to print output to
 */
static FILE * outfile;

/**
 * secs:    number of seconds to offset the clip start/end times
 */
static double secs;

/**
 * PrintUsage: prints out help on how to use this program
 *
 * \param prog the program's name
 */
static void
PrintUsage(char *prog) {
  fprintf(stderr, "Usage: %s [options] filename\n", prog);
  fprintf(stderr, "Apply a negative or positive time offset (in seconds) to a cmml file.\n\n");
  fprintf(stderr, "Possible options:\n");
#ifdef HAVE_GETOPT_LONG
  fprintf(stderr, "  -s offset, --sec offset\n");
  fprintf(stderr, "                 Specify the offset in seconds to add to clips\n");
  fprintf(stderr, "  -o filename, --output filename\n");
  fprintf(stderr, "                 Specify the output filename. The file is written\n");
  fprintf(stderr, "                 to standard output by default.\n");
  fprintf(stderr, "  -h, --help     Display this help information\n");
  fprintf(stderr, "  -v, --version  Display version information\n");
#else
  fprintf(stderr, "  -s offset\n");
  fprintf(stderr, "                 Specify the offset in seconds to add to clips\n");
  fprintf(stderr, "  -o filename\n");
  fprintf(stderr, "                 Specify the output filename. The file is written\n");
  fprintf(stderr, "                 to standard output by default.\n");
  fprintf(stderr, "  -h             Display this help information\n");
  fprintf(stderr, "  -v             Display version information\n");
#endif
  fprintf(stderr, "\nPlease report bugs to <libcmml-devel@cmis.csiro.au>.\n");
  exit(1);
}

/**
 * read_stream: the callback for a stream element
 *
 * \param cmml the CMML* handle in use
 * \param stream the stream element's content represented in a
 *        CMML_Stream*
 * \param user_data user defined data
 *
 * \returns 0 on success, 1 on error
 */
static int
read_stream (CMML * cmml, const CMML_Stream * stream, void * user_data) {
  char buf[BUFSIZE];
  CMML_Error * err;
  if ((err = cmml_get_last_error(cmml)) != NULL) {
    cmml_error_snprint(buf, BUFSIZE, err, cmml);
    fprintf(stderr, "cmml-timeshift: Parsing stream tag %s\n", buf);
    fprintf(stderr, "cmml-timeshift: Non-recoverable error\n");
    return -1;
  } else {
    /* print stream */
    cmml_stream_pretty_snprint (buf, BUFSIZE, (CMML_Stream *) stream);
    fprintf(outfile, "%s\n", buf);
  }
  return 0;
}

/**
 * read_head: the callback for a head element
 *
 * \param cmml the CMML* handle in use
 * \param head the head element's content represented in a
 *        CMML_Head*
 * \param user_data user defined data
 *
 * \returns 0 on success, 1 on error
 */
static int
read_head (CMML * cmml, const CMML_Head * head, void * user_data) {
  char buf[BUFSIZE];
  CMML_Error * err;
  if ((err = cmml_get_last_error(cmml)) != NULL) {
    cmml_error_snprint(buf, BUFSIZE, err, cmml);
    fprintf(stderr, "cmml-timeshift: Parsing head tag %s\n", buf);
    fprintf(stderr, "cmml-timeshift: Non-recoverable error\n");
    return -1;
  } else {
    cmml_head_pretty_snprint (buf, BUFSIZE, (CMML_Head *) head);
    fprintf(outfile, "%s\n", buf);
  }
  return 0;
}

/**
 * read_clip: the callback for a clip element
 *
 * \param cmml the CMML* handle in use
 * \param clip the clip element's content represented in a
 *        CMML_Clip*
 * \param user_data user defined data
 *
 * \returns 0 on success, 1 on error
 */
static int
read_clip (CMML * cmml, const CMML_Clip * clip, void * user_data) {
  char buf[BUFSIZE];
  CMML_Error * err;
  CMML_Clip * lclip;
  double newsecs;

  if ((err = cmml_get_last_error(cmml)) != NULL) {
    cmml_error_snprint(buf, BUFSIZE, err, cmml);
    fprintf(stderr, "cmml-timeshift: Parsing clip %s\n", buf);
    fprintf(stderr, "cmml-timeshift: Skipping clip\n\n");
    return -1;
  } else {
    lclip = cmml_clip_clone((CMML_Clip *)clip);

    /* shift clip's start time by secs */
    newsecs = cmml_sec_parse(lclip->start_time->tstr) + secs;
    cmml_time_free(lclip->start_time);
    lclip->start_time = cmml_time_new_secs (newsecs);

    /* check end time, too */
    if (lclip->end_time != NULL) {
      newsecs = cmml_sec_parse(lclip->end_time->tstr) + secs;
      cmml_time_free(lclip->end_time);
      lclip->end_time = cmml_time_new_secs (newsecs);
    }

    /* print clip if beyond start time */
    if (lclip->start_time->t.sec >= 0.0) {
      cmml_clip_pretty_snprint (buf, BUFSIZE, lclip);
      fprintf(outfile, "%s\n", buf);
    }

    cmml_clip_destroy(lclip);
  }

  return 0;
}

/**
 * main function of cmml-timeshift, which opens the CMML file, seeks to
 * any given offsets, registers the callbacks, and then steps through
 * the file in chunks of BUFSIZE size, during which the callbacks get
 * activated as the relevant elements get parsed.
 */
int main(int argc, char *argv[])
{
  char *pathfile = NULL;
  int i;
  char buf[BUFSIZE];
  CMML * doc;
  CMML_Error * err;
  CMML_Preamble * pre;
  long n = 0;
  char * outfilename = NULL;
  buf[0]='\0';

  /* default outfile to stdout and secs to 0 */
  outfile = stdout;
  secs = 0;

  while (1) {
    char * optstring = "hvos:";

#ifdef HAVE_GETOPT_LONG
    static struct option long_options[] = {
      {"help",no_argument,0,'h'},
      {"version",no_argument,0,'v'},
      {"output", required_argument,0,'o'},
      {"offset", required_argument,0,'s'},
      {0,0,0,0}
    };

    i = getopt_long(argc, argv, optstring, long_options, NULL);
#else
    i = getopt(argc, argv, optstring);
#endif

    if (i == -1) break;
    if (i == ':') PrintUsage(argv[0]);

    switch (i) {
    case 'h': /* help */
      PrintUsage(argv[0]);
      break;
    case 'v': /* version */
      fprintf(stdout, "cmml-timeshift version " VERSION "\n");
      fprintf(stdout, "# cmml-timeshift, Copyright (C) 2005 CSIRO Australia www.csiro.au ; www.annodex.net\n");
      break;
    case 'o': /* output */
      outfilename = optarg;
      break;
    case 's': /* offset */
      /* parse seconds specification */
      secs = atof(optarg);
      break;
    default:
      break;
    }
  }

  /* more arguments that were not parsed */
  if (optind > argc) {
    PrintUsage(argv[0]);
  }

  /* open output file */
  if (outfilename != NULL) {
    if ((outfile = fopen (outfilename, "wb")) == NULL) {
      fprintf (stderr, "%s: %s: Error opening %s for writing\n", 
	       argv[0], pathfile, outfilename);
      exit (1);
    }
  }

  /* no filename given? */
  if (optind == argc) {
    pathfile = "-";
  } else {
    pathfile = argv[optind++];
  }

  /* try open file for parsing and setup cmml parsing */
  errno=0;
  if (strcmp (pathfile, "-") == 0) {
    doc = cmml_new (stdin);
  } else {
    doc = cmml_open (pathfile);
  }

  if (doc == NULL) {
    if (errno == 0) {
      fprintf(stderr, "%s: %s: CMML error opening file\n", argv[0], pathfile);
    } else {
      fprintf(stderr, "%s: %s: %s\n", argv[0], pathfile, strerror(errno));
    }
    PrintUsage(argv[0]);
  }

  /* turn on sloppy parsing */
  cmml_set_sloppy(doc, 1);

  /* print preamble */
  pre = cmml_get_preamble(doc);
  cmml_preamble_snprint(buf, BUFSIZE, pre);
  fprintf(outfile, "%s\n", buf);

  /* register callbacks */
  cmml_set_read_callbacks (doc, read_stream, read_head, read_clip, NULL);

  /* read document frame-wise and check against CMML.dtd */
  while ((n = cmml_read (doc, BUFSIZE)) > 0) {
    /* if error reading, print and exit */
    if ((err = cmml_get_last_error(doc)) != NULL && err->type != CMML_EOF) {
      char *filename;
      filename = (strrchr(pathfile, '/') == NULL ? pathfile
		  : strrchr(pathfile, '/')+1);
      cmml_error_snprint(buf, BUFSIZE, err, doc);
      fprintf (stderr, "%s:%s\n", filename, buf);
      goto cleanup;
    }
  }

  err = cmml_get_last_error(doc);
  if (err!=NULL && err->type != CMML_EOF) {
      char *filename;
      filename = (strrchr(pathfile, '/') == NULL ? pathfile
		  : strrchr(pathfile, '/')+1);
      cmml_error_snprint(buf, BUFSIZE, err, doc);
      fprintf (stderr, "%s:%s\n", filename, buf);
      goto cleanup;
  }

  /* write end tag */
  fprintf(outfile, "</cmml>\n");
  
 cleanup:
  /* clean up */
  cmml_close(doc);

  return 0;
}
